// sheweny/build/structures/Event.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Event = void 0;
const collection_data_1 = require("collection-data");
const _1 = require(".");
/**
 * Represents an Event structure
 * @extends {BaseStructure}
 */
class Event extends _1.BaseStructure {
    /**
     * Constructor for build a Event
     * @param {ShewenyClient} client Client framework
     * @param {keyof ClientEvents} name Name of the event
     * @param {string[]} customId Custom id for one or more buttons
     */
    constructor(client, name, options) {
        super(client);
        this.client = client;
        this.name = name;
        this.description = options?.description || "";
        this.once = options?.once || false;
    }
    /**
     * Unregister a event
     * @public
     * @returns {boolean}
     */
    unregister() {
        this.client.collections.events?.delete(this.name);
        delete require.cache[require.resolve(this.path)];
        return true;
    }
    /**
     * Reload a event
     * @public
     * @async
     * @returns {Promise<Collection<string, Event> | null>} The events collection
     */
    async reload() {
        if (this.path) {
            this.unregister();
            return this.register();
        }
        return null;
    }
    /**
     * Register a event
     * @public
     * @async
     * @returns {Promise<Collection<string, Event>>} The events collection
     */
    async register() {
        const event = (await Promise.resolve().then(() => require(this.path))).default;
        const evt = new event(this.client);
        return this.client.collections.events
            ? this.client.collections.events.set(evt.name, evt)
            : new collection_data_1.Collection().set(evt.name, evt);
    }
}
exports.Event = Event;
