// sheweny/build/structures/Inhibitor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Inhibitor = void 0;
const collection_data_1 = require("collection-data");
const _1 = require(".");
/**
 * Represents an Command structure
 * @extends {BaseStructure}
 */
class Inhibitor extends _1.BaseStructure {
    /**
     * Constructor for build a Inhibitor
     * @param {ShewenyClient} client Client framework
     * @param {string} name Name of the event
     * @param {InhibitorOptions} [options] Options for the inhibitor
     */
    constructor(client, name, options) {
        super(client);
        this.client = client;
        this.name = name;
        this.type = options?.type || ["MESSAGE_COMMAND"];
        this.priority = options?.priority || 0;
    }
    /**
     * Unregister a inhibitor from collections
     * @returns {boolean}
     */
    unregister() {
        this.client.collections.inhibitors?.delete(this.name);
        delete require.cache[require.resolve(this.path)];
        return true;
    }
    /**
     * Reload a inhibitor
     * @returns {Promise<Collection<string[], Inhibitor> | null>} The inhibitors collection
     */
    async reload() {
        if (this.path) {
            this.unregister();
            return this.register();
        }
        return null;
    }
    /**
     * Register a inhibitor in collections
     * @returns {Collection<string[], Inhibitor>} The inhibitors collection
     */
    async register() {
        const Inhibitor = (await Promise.resolve().then(() => require(this.path))).default;
        const inhib = new Inhibitor(this.client);
        return this.client.collections.inhibitors
            ? this.client.collections.inhibitors.set(inhib.name, inhib)
            : new collection_data_1.Collection().set(inhib.name, inhib);
    }
}
exports.Inhibitor = Inhibitor;
